### PR TITLE
[WIP] Audit trail

### PR DESCRIPTION
**Pull request by stevenbal**

### CODE REVIEW

File: src/zrc/api/audittrails.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
+from ..conf.base import SITE_TITLE
+from ..datamodel.models import AuditTrail
+
+
+class AuditTrailMixin:
+    def create_audittrail(self, status_code, action, oud, nieuw):
+        data = nieuw if nieuw else oud
+        if SITE_TITLE == 'Zaak Registratie Component (ZRC)':
+            bron = 'ZRC'
+            hoofdObject = data['url'] if self.basename == 'zaak' else data['zaak']
```
Comment (joeribekker) [inline src/zrc/api/audittrails.py]: This needs to work similar to how notifications decides on the `hoofdObject`.

File: src/zrc/api/audittrails.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
+from django.db import transaction
+
+from ..conf.base import SITE_TITLE
```
Comment (joeribekker) [inline src/zrc/api/audittrails.py]: cleanup

File: src/zrc/api/audittrails.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+from ..datamodel.models import AuditTrail
+
+
+class AuditTrailMixin:
+    bron = None
+    main_resource = None
```
Comment (joeribekker) [inline src/zrc/api/audittrails.py]: Let's define an `Audit` class just like `Scope` and `Kanaal` objects. In it, we can store "main_resource" and "component_name" (bron).

File: src/zrc/api/audittrails.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+            resultaat=status_code,
+            hoofdObject=main_object,
+            resource=self.basename,
+            resourceUrl=data['url'],
+            oud=version_before_edit,
+            nieuw=version_after_edit,
+        )
+        trail.save()
+
+    def destroy_related_audittrails(self, main_object_url):
```
Comment (joeribekker) [inline src/zrc/api/audittrails.py]: Why is this here, and not in `AuditTrailDestroyMixin`?

File: src/zrc/api/audittrails.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+            'create',
+            version_before_edit=None,
+            version_after_edit=response.data,
+        )
+        return response
+
+
+class AuditTrailUpdateMixin(AuditTrailMixin):
+    def update(self, request, *args, **kwargs):
+        version_before_edit = self.get(request, *args, **kwargs).data
```
Comment (joeribekker) [inline src/zrc/api/audittrails.py]: This is quite high level, I suggest something like:

```python
        instance = self.get_object()
        serializer = self.get_serializer(instance)
        version_before_edit = serializer.data
```

File: src/zrc/api/audittrails.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        action = 'update' if request.method == 'PUT' else 'partial_update'
```
Comment (joeribekker) [inline src/zrc/api/audittrails.py]:
```python
action = 'partial_update' if kwargs.get('partial', False) else 'update'
```

File: src/zrc/api/audittrails.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+            action,
+            version_before_edit=version_before_edit,
+            version_after_edit=response.data,
+        )
+        return response
+
+
+class AuditTrailDestroyMixin(AuditTrailMixin):
+    def destroy(self, request, *args, **kwargs):
+        version_before_edit = self.get(request, *args, **kwargs).data
```
Comment (joeribekker) [inline src/zrc/api/audittrails.py]: See above.

File: src/zrc/api/audittrails.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+        if self.basename == self.main_resource:
+            with transaction.atomic():
+                response = super().destroy(request, *args, **kwargs)
+                self.destroy_related_audittrails(version_before_edit['url'])
```
Comment (joeribekker) [inline src/zrc/api/audittrails.py]: can be a protected function (start with `_`).

File: src/zrc/api/audittrails.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+                return response
+        else:
+            response = super().destroy(request, *args, **kwargs)
+            self.create_audittrail(
+                response.status_code,
+                'delete',
```
Comment (joeribekker) [inline src/zrc/api/audittrails.py]: If you move this to VNG-API-Common, please make a DjangoChoices with all the actions and use these for notifications and audittrail. Note that the model fields should still allow strings because custom actions can occur but let's at least define constants for the basics (update, get, delete, partial_update, etc.).

File: src/zrc/api/serializers.py [hunk 10]
```diff
@@ -622,3 +622,26 @@ class Meta:
...
+        model = AuditTrail
+        fields = (
+            'uuid',
+            'bron',
+            # 'applicatieId',
+            # 'applicatieWeergave',
+            # 'gebruikersId',
+            # 'gebruikersWeergave',
+            'actie',
+            'actieWeergave',
```
Comment (joeribekker) [inline src/zrc/api/serializers.py]: We can store the (above define) DjangoChoices representation of the constant for `actieWeergave` by default.

File: src/zrc/api/tests/test_audittrails.py [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+            'startdatum': '2018-12-24',
+            'productenOfDiensten': ['https://example.com/product/123']
+        }
+        with mock_client(self.responses):
+            response = self.client.post(url, zaak_data, **ZAAK_WRITE_KWARGS)
+
+        return response.data
+
+    def test_create_zaak_audittrail(self):
+        zaak_response = self.create_zaak()
```
Comment (joeribekker) [inline src/zrc/api/tests/test_audittrails.py]: Use a factory instead?

File: src/zrc/api/viewsets.py [hunk 12]
```diff
@@ -165,6 +171,9 @@ class ZaakViewSet(NotificationViewSetMixin,
     }
     notifications_kanaal = KANAAL_ZAKEN
 
+    bron = 'ZRC'
+    main_resource = 'zaak'
```
Comment (joeribekker) [inline src/zrc/api/viewsets.py]: both variables: no.

File: src/zrc/api/viewsets.py [hunk 13]
```diff
@@ -506,3 +544,19 @@ class ResultaatViewSet(NotificationViewSetMixin,
...
+
+
+class ZaakAuditTrailViewset(viewsets.ReadOnlyModelViewSet, NestedViewSetMixin):
+    queryset = AuditTrail.objects.all()
+    serializer_class = ZaakAuditTrailSerializer
+    lookup_field = 'uuid'
+
+    def get_queryset(self):
+        base = super().get_queryset()
+        zaak_uuid = self.request.parser_context['kwargs'].get('zaak_uuid')
```
Comment (joeribekker) [inline src/zrc/api/viewsets.py]: I think you can get it easier with `self.kwargs.get('zaak_uuid')`

File: src/zrc/datamodel/migrations/0057_audittrail.py [hunk 14]
```diff
@@ -0,0 +1,29 @@
+# Generated by Django 2.0.13 on 2019-05-02 15:22
```
Comment (joeribekker) [inline src/zrc/datamodel/migrations/0057_audittrail.py]: Squash migrations please (delete the new ones in your branch, and create just one)

File: src/zrc/datamodel/models.py [hunk 15]
```diff
@@ -489,3 +490,25 @@ def save(self, *args, **kwargs):
...
         super().save(*args, **kwargs)
+
+
+class AuditTrail(models.Model):
+    uuid = models.UUIDField(
+        unique=True, default=uuid.uuid4,
+        help_text="Unieke resource identifier (UUID4)"
+    )
+    bron = models.CharField(max_length=3)
+    actie = models.CharField(max_length=20)
```
Comment (joeribekker) [inline src/zrc/datamodel/models.py]: max_length=50

File: src/zrc/datamodel/models.py [hunk 16]
```diff
@@ -489,3 +490,25 @@ def save(self, *args, **kwargs):
...
             self.identificatie = identificatie
         super().save(*args, **kwargs)
+
+
+class AuditTrail(models.Model):
+    uuid = models.UUIDField(
+        unique=True, default=uuid.uuid4,
+        help_text="Unieke resource identifier (UUID4)"
+    )
+    bron = models.CharField(max_length=3)
```
Comment (joeribekker) [inline src/zrc/datamodel/models.py]: max_length=50

File: src/zrc/datamodel/models.py [hunk 17]
```diff
@@ -489,3 +490,25 @@ def save(self, *args, **kwargs):
...
+    actie = models.CharField(max_length=20)
+    actieWeergave = models.CharField(max_length=30, blank=True)
```
Comment (joeribekker) [inline src/zrc/datamodel/models.py]: max_length=200

File: src/zrc/datamodel/models.py [hunk 18]
```diff
@@ -489,3 +490,25 @@ def save(self, *args, **kwargs):
...
+    resultaat = models.IntegerField()
+    hoofdObject = models.URLField()
+    resource = models.CharField(max_length=20)
```
Comment (joeribekker) [inline src/zrc/datamodel/models.py]: max_length=50

File: src/zrc/api/tests/test_audittrails.py (hunk 11 above)
Comment (stevenbal) [reply src/zrc/api/tests/test_audittrails.py]: But, if I understand it correctly, using a factory to create a Zaak would not use the API to do so, which means that no Audittrail will be created for that operation, right?

Comment (joeribekker) [reply src/zrc/api/tests/test_audittrails.py]: Woops, correct :)

**Comment by joeribekker** [comment]
I see all feedback was processed in vng-api-common: https://github.com/VNG-Realisatie/vng-api-common/pull/24

And the ZRC specific is now in PR #57 

Closing this PR.
